### PR TITLE
feat: Support commit delay

### PR DESCRIPTION
**Pull request by amanda-tarafa**
@jskeet As always one commit at a time.

And happy for alternatives, in particular for the third commit.

FYI: @nginsberg-google

### CODE REVIEW

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerBatchCommandTests.cs [hunk 1]
```diff
@@ -212,6 +213,77 @@ public void EphemeralTransactionIncludesPriorityOnBatchDmlAndCommit()
...
+            command.CommitDelay = commitDelay;
+            command.ExecuteNonQuery();
+
+            spannerClientMock.Received(1).CommitAsync(
+                Arg.Is<CommitRequest>(request => request.MaxCommitDelay.Equals(Duration.FromTimeSpan(commitDelay))),
+                Arg.Any<CallSettings>());
+        }
+
+        [Fact]
+        public void CommitDelay_SetOnCommand_IgnoredOnExplicitTransaction()
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerBatchCommandTests.cs:261]: What about if it's set on the command, but not set on the transaction? (It's not obvious from the tests if this is meant to be inherited with overriding at the transaction, or "only whatever's on the transaction". The docs seem to suggest it's "only whatever's on the transaction" but I think that's worth a test.)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs [hunk 2]
```diff
@@ -834,6 +834,73 @@ public void CommitDelay_Propagates_RunWithRetryableTransaction()
...
+            command.CommitDelay = commitDelay;
+            command.ExecuteNonQuery();
+
+            spannerClientMock.Received(1).CommitAsync(
+                Arg.Is<CommitRequest>(request => request.MaxCommitDelay.Equals(Duration.FromTimeSpan(commitDelay))),
+                Arg.Any<CallSettings>());
+        }
+
+        [Fact]
+        public void CommitDelay_SetOnCommand_IgnoredOnExplicitTransaction()
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs:877]: (Same here.)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommand.cs [hunk 3]
```diff
@@ -281,6 +281,22 @@ public override UpdateRowSource UpdatedRowSource
         /// </remarks>
         public DirectedReadOptions DirectedReadOptions { get; set; }
 
+        /// <summary>
+        /// The maximum amount of time the commit of the implicit transaction associated with this command, if any,
+        /// may be delayed server side for batching with other commits.
+        /// The bigger the delay, the better the throughput (QPS), but at the expense of commit latency.
+        /// If set to <see cref="TimeSpan.Zero"/>, commit batching is disabled.
+        /// May be null, in which case commits will continue to be batched as they had been before this configuration
+        /// option was made available to Spaner API consumers.
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommand.cs:290]: Spaner => Spanner

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommand.cs [hunk 4]
```diff
@@ -281,6 +281,22 @@ public override UpdateRowSource UpdatedRowSource
...
+        /// May be set to any value between <see cref="TimeSpan.Zero"/> and 500ms.
+        /// </summary>
+        /// <remarks>
+        /// When a DML or mutation command is executed with no explicit or ambient transaction, an implicit transaction is created
+        /// and the command is executed within it. This value will be applied to the commit operation of such transaction,
+        /// if there is any. Otherwise, this value will be ignored.
+        /// </remarks>
+        public TimeSpan? CommitDelay { get; set; }
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommand.cs:298]: Again, possible validation.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs [hunk 5]
```diff
@@ -901,6 +901,84 @@ public void CommitDelay_SetOnCommand_IgnoredOnExplicitTransaction()
...
+                scope.Complete();
+            }
+
+            spannerClientMock.Received(1).CommitAsync(
+                Arg.Is<CommitRequest>(request => request.MaxCommitDelay.Equals(Duration.FromTimeSpan(commitDelay))),
+                Arg.Any<CallSettings>());
+        }
+
+        [Fact]
+        public void CommitDelay_SetOnCommand_IgnoredOnAmbientTransaction()
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs:955]: Same comment as before: another test showing that if the ambient transaction doesn't have a commit delay, we don't use it.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs [hunk 6]
```diff
@@ -943,19 +943,59 @@ public Task OpenAsReadOnlyAsync(TimestampBound timestampBound = null, Cancellati
...
+            Action transactionEnlister = () => EnlistTransaction(transaction, timestampBound ?? TimestampBound.Strong, transactionId: null, commitdelay: null);
             return OpenAsyncImpl(transactionEnlister, cancellationToken);
         }
 
+        /// <summary>
+        /// Opens the connection within a <see cref="System.Transactions.TransactionScope"/> specifying a commit delay
+        /// value for the associated Cloud Spanner transaction. See <see cref="SpannerTransaction.CommitDelay"/> for
+        /// more information.
+        /// </summary>
+        public void Open(TimeSpan commitDelay)
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:955]: I'm slightly nervous that commit delay might only be *one* option that we want to specify like this, and we could end up with a load of overloads in the future. Would some sort of Spanner.Data-level TransactionOptions type be worth introducing?

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs [hunk 7]
```diff
@@ -943,19 +943,59 @@ public Task OpenAsReadOnlyAsync(TimestampBound timestampBound = null, Cancellati
...
         /// Gets or Sets whether to participate in the active <see cref="System.Transactions.TransactionScope" />
         /// </summary>
         public bool EnlistInTransaction { get; set; } = true;
 
         /// <inheritdoc />
-        public override void EnlistTransaction(Transaction transaction) => EnlistTransaction(transaction, null, null);
+        public override void EnlistTransaction(Transaction transaction) => EnlistTransaction(transaction, null, null, null);
 
-        private void EnlistTransaction(Transaction transaction, TimestampBound timestampBound, TransactionId transactionId)
+        private void EnlistTransaction(Transaction transaction, TimestampBound timestampBound, TransactionId transactionId, TimeSpan? commitdelay)
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:998]: Nit: commitdelay => commitDelay

**Review by jskeet — COMMENTED** [review]
Some nits and a few relevant suggestions. Happy to chat about any decisions e.g. validation.

**Review by amanda-tarafa — COMMENTED** [review]
@jskeet I've addressed all your comments in ordered review commits. I've also added a new commit, the last one, that uses the newly introduced AmbientTransactionOptions for creation of all parameterized ambient transactions. PTAL, thanks!

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs [hunk 8]
```diff
@@ -563,5 +570,10 @@ private void CheckCompatibleMode(TransactionMode mode)
                     throw new ArgumentOutOfRangeException(nameof(mode), mode, null);
             }
         }
+
+        private static readonly TimeSpan s_maxCommitDelay = TimeSpan.FromMilliseconds(500);
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs:574]: I'm somewhat nervous of building this into the client. I think it's absolutely fine to check that it's not negative, but I can easily see the Spanner team increasing this over time, and the libraries getting out of date.

Effectively, I'd say that "not negative" is a *logical* validation, whereas "at most 500ms" is an *implementation-specific* validation.

Happy to be persuaded though :)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs [hunk 9]
```diff
@@ -889,41 +889,18 @@ private Action GetTransactionEnlister()
...
-            {
-                throw new InvalidOperationException($"{nameof(OpenAsReadOnlyAsync)} should only be called within a TransactionScope.");
-            }
-            if (!EnlistInTransaction)
-            {
-                throw new InvalidOperationException($"{nameof(OpenAsReadOnlyAsync)} should only be called with ${nameof(EnlistInTransaction)} set to true.");
-            }
-            Open(() => EnlistTransaction(transaction, timestampBound ?? TimestampBound.Strong, transactionId: null, commitDelay: null));
-        }
+        [Obsolete($"Use the {nameof(Open)} that takes a {nameof(AmbientTransactionOptions)} parameter instead.")]
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:892]: I like the new methods... I'm not sure about making the existing ones obsolete immediately. (I'm really on the fence.) Let's chat about it.

**Review by amanda-tarafa — COMMENTED** [review]
There's a new commit prefixed `review:` with the validation changes. And I cna change deprecation of the existing ambient transaction related methods after we've chatted about it.

**Comment by amanda-tarafa** [comment]
We discussed internally and deprecating the existing ambien transaction methods makes is OK, given that we are releasing a new major version.

@jskeet Can you stamp?

**Review by jskeet — APPROVED** [review]

**Pull request merged**